device.go

```diff
@@ -41,8 +62,51 @@ func blockDeviceHandler(device pb.Device, spec *pb.Spec) error {
 		return nil
 	}
 
-	// At this point in the code, we assume the specification will be
-	// updated, meaning we should make sure we have valid pointers here.
+	return updateSpecDeviceList(device, spec)
+}
+
+func scsiBlockDeviceHandler(device pb.Device, mounts []*pb.Storage, spec *pb.Spec) error {
+	// Retrieve the device path from SCSI address.
+	devPath, err := getSCSIDevPath(device.Id)
+	if err != nil {
+		return err
+	}
+	device.VmPath = devPath
+
+	// If no ContainerPath is provided, this means we don't expect the
+	// device to be updated in the specification. Instead, this means
+	// this device is the rootfs, and the mount list should be updated
+	// accordingly.
+	if device.ContainerPath == "" {
+		return updateRootfsMountList(device, mounts)
+	} else {
+		return updateSpecDeviceList(device, spec)
+	}
+}
+
+func updateRootfsMountList(device pb.Device, mounts []*pb.Storage) error {
+	if len(mounts) < 1 {
+		return grpcStatus.Errorf(codes.Internal, "Mount list is empty")
+	}
+
+	for _, mnt := range mounts {
+		if mnt == nil {
+			continue
+		}
+
+		// An empty source here means this is the rootfs and it can
+		// be replaced with the provided value of the device VM path.
```

Comment:
I agree the empty `Source` is not really clean here, and I think using a field `RootFs bool` into `Storage` structure could be easier to understand.

Reply:
`Storage` has a `MountPoint` which defined by the runtime and it will be located in the VM.
After `Storage` handled in the agent, all the dir/files of MountPoints can be used and all the path (including oci.Root.Path & oci.Mount.Srouce) in the ocispec can use these paths(Storage.MountPoint). `RootFs bool` is needless.